src/main/drivers/rx/rx_sx127x.c

```diff
@@ -0,0 +1,471 @@
+        sx127xSetRegisterValue(SX127X_REG_DETECT_OPTIMIZE, SX127X_DETECT_OPTIMIZE_SF_7_12, 2, 0);
+        sx127xWriteRegister(SX127X_REG_DETECTION_THRESHOLD, SX127X_DETECTION_THRESHOLD_SF_7_12);
+    }
+}
+
+void sx127xSetFrequencyHZ(const uint32_t freq)
+{
+    sx127xSetMode(SX127x_OPMODE_STANDBY);
+
+    int32_t FRQ = ((uint32_t)((double)freq / (double)SX127x_FREQ_STEP));
+
+    uint8_t FRQ_MSB = (uint8_t)((FRQ >> 16) & 0xFF);
+    uint8_t FRQ_MID = (uint8_t)((FRQ >> 8) & 0xFF);
+    uint8_t FRQ_LSB = (uint8_t)(FRQ & 0xFF);
+
+    uint8_t outbuff[3] = {FRQ_MSB, FRQ_MID, FRQ_LSB};
+
+    sx127xWriteRegisterBurst(SX127X_REG_FRF_MSB, outbuff, sizeof(outbuff));
+}
+
+void sx127xSetFrequencyReg(const uint32_t freq)
+{
+    sx127xSetMode(SX127x_OPMODE_STANDBY);
+
+    uint8_t FRQ_MSB = (uint8_t)((freq >> 16) & 0xFF);
+    uint8_t FRQ_MID = (uint8_t)((freq >> 8) & 0xFF);
+    uint8_t FRQ_LSB = (uint8_t)(freq & 0xFF);
+
+    uint8_t outbuff[3] = {FRQ_MSB, FRQ_MID, FRQ_LSB}; //check speedup
+
+    sx127xWriteRegisterBurst(SX127X_REG_FRF_MSB, outbuff, sizeof(outbuff));
+}
+
+void sx127xTransmitData(const uint8_t *data, const uint8_t length)
+{
+    sx127xSetMode(SX127x_OPMODE_STANDBY);
+
+    sx127xWriteRegister(SX127X_REG_FIFO_ADDR_PTR, SX127X_FIFO_TX_BASE_ADDR_MAX);
+    sx127xWriteRegisterFIFO(data, length);
+
+    sx127xSetMode(SX127x_OPMODE_TX);
+}
+
+void sx127xReceiveData(uint8_t *data, const uint8_t length)
+{
+    sx127xReadRegisterFIFO(data, length);
+}
+
+void sx127xStartReceiving(void)
+{
+    sx127xSetMode(SX127x_OPMODE_STANDBY);
+    sx127xWriteRegister(SX127X_REG_FIFO_ADDR_PTR, SX127X_FIFO_RX_BASE_ADDR_MAX);
+    sx127xSetMode(SX127x_OPMODE_RXCONTINUOUS);
+}
+
+void sx127xConfig(const sx127x_bandwidth_e bw, const sx127x_spreading_factor_e sf, const sx127x_coding_rate_e cr, const uint32_t freq, const uint8_t preambleLen, const bool iqInverted)
+{
+    sx127xConfigLoraDefaults(iqInverted);
+    sx127xSetPreambleLength(preambleLen);
+    sx127xSetOutputPower(SX127x_MAX_POWER);
+    sx127xSetSpreadingFactor(sf);
+    sx127xSetBandwidthCodingRate(bw, cr, sf, false, false);
+    sx127xSetFrequencyReg(freq);
+}
+
+uint32_t sx127xGetCurrBandwidth(const sx127x_bandwidth_e bw)
+{
+    switch (bw) {
+    case SX127x_BW_7_80_KHZ:
+        return 7.8E3;
+    case SX127x_BW_10_40_KHZ:
+        return 10.4E3;
+    case SX127x_BW_15_60_KHZ:
+        return 15.6E3;
+    case SX127x_BW_20_80_KHZ:
+        return 20.8E3;
+    case SX127x_BW_31_25_KHZ:
+        return 31.25E3;
+    case SX127x_BW_41_70_KHZ:
+        return 41.7E3;
+    case SX127x_BW_62_50_KHZ:
+        return 62.5E3;
+    case SX127x_BW_125_00_KHZ:
+        return 125E3;
+    case SX127x_BW_250_00_KHZ:
+        return 250E3;
+    case SX127x_BW_500_00_KHZ:
+        return 500E3;
+    }
+    return -1;
+}
+
+uint32_t sx127xGetCurrBandwidthNormalisedShifted(const sx127x_bandwidth_e bw) // this is basically just used for speedier calc of the freq offset, pre compiled for 32mhz xtal
```

Comment:
Can we have the comment above the function to make it easier to read?

Reply:
put the comment above